### PR TITLE
Brian Flynn's Siiiiiiick Weather App

### PR DESCRIPTION
**Pull request by briandangerflynn**
Hey Contently,

This was a fun challenge; thanks for the opportunity! Looking forward to hearing back.

Much appreciated,

Brian

### CODE REVIEW

File: src/components/Header.jsx [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    handleView,
+    handleClear
+  } = props
+
+  return (
+    <header>
+      <h1><Logo />Weather</h1>
+      <div>
+        <input required type="text" name="cityName" onChange={handleChange} onClick={handleClear} placeholder="Enter City... " />
+        <button onClick={handleClick}>Get Weather</button>
```
Comment (jackpope) [inline src/components/Header.jsx:21]: What are your thoughts on using the components with click handlers vs using a `<form>` submission flow?

File: src/components/app.jsx [hunk 2]
```diff
@@ -1,9 +1,163 @@
...
 import "../App.css";
+import { getWeather, getForecast } from '../utils.js';
+import Header from './Header';
+import WeatherCurrent from './WeatherCurrent';
+import WeatherForecast from './WeatherForecast';
 
-function App() {
-  return <h1>Weather</h1>;
+class App extends React.Component {
+  state = {
```
Comment (jackpope) [inline src/components/app.jsx:9]: Holding all of the state and handling mutations at the App root level is workable for a simple app like this. If we had to scale out the feature-set here, what strategies might you consider for refactoring the data flow?

File: src/index.js [hunk 3]
```diff
@@ -1,5 +1,5 @@
 import React from "react";
 import ReactDOM from "react-dom";
-import App from "./components/app";
+import App from "./components/App";
```
Comment (jackpope) [inline src/index.js:3]: Curious why this change was needed? I had to change it back in order to launch the app locally.

File: src/components/app.jsx [hunk 4]
```diff
@@ -1,9 +1,163 @@
...
-import { ReactComponent as Logo } from "../images/sun.svg";
 import "../App.css";
+import { getWeather, getForecast } from '../utils.js';
+import Header from './Header';
+import WeatherCurrent from './WeatherCurrent';
+import WeatherForecast from './WeatherForecast';
 
-function App() {
-  return <h1>Weather</h1>;
+class App extends React.Component {
```
Comment (jackpope) [inline src/components/app.jsx:8]: Have you worked with React hooks before? If so, how do you decide to make class-based components vs functional components with hooks?

File: src/components/app.jsx [hunk 5]
```diff
@@ -1,9 +1,163 @@
...
+      this.setState({
+        cityName
+      })
+      this.handleWeatherState(cityName, unit)
+      this.handleForecastState(cityName, unit)
+    }
+  }
+
+  handleWeatherState = async (cityName, unit) => {
+    const response = await getWeather(cityName, unit)
```
Comment (jackpope) [inline src/components/app.jsx:36]: What would happen if there was a network error?

**Review by jackpope — COMMENTED** [review]
Hi Brian!

I'm a software engineer here at Contently and I'll be reviewing your submission. Thanks for the good work! I especially like the header you made with toggle controls for temp format and current/forecast results which all feel very intuitive.

I've left some questions on the code that I'd love to hear your opinions on as you have the time.

File: src/components/Header.jsx (hunk 1 above)
Comment (briandangerflynn) [reply src/components/Header.jsx:21]: Normally, I would wrap everything in a form and use `onSubmit`. I almost did that here, again; however, since it's only one input and a simple `get` request, it just didn't feel necessary. That said, when I was done with the app, I kind of regretted this, as I realized it would have been nice if you could click the button OR hit `enter`, and an `onSubmit` would have made that easier. If this was a real-life project, I probably would have updated it.

File: src/components/app.jsx (hunk 2 above)
Comment (briandangerflynn) [reply src/components/app.jsx:9]: Typically, on larger apps, if a child component needs to be stateful, I consider whether the data is necessary for just for that component. If it is, then I store that data in the child component's state instead of in the root component's state. For example, I usually make `<Login />` components stateful, as the data and methods needed there aren't really necessary in the rest of the app. To my understanding, this process could be even further refactored with React Redux or Context API, as they should be able to separate the state management from the presentation logic; however, I don't currently have experience with either of those (would love to learn though!).

File: src/index.js (hunk 3 above)
Comment (briandangerflynn) [reply src/index.js:3]: At General Assembly, our style guide has us use PascalCase for every component, even App.jsx, so I changed the file name out of habit. I've never had an issue with this before! It worked perfectly fine on my computer locally. Glad you were able to get it up and running, and sorry such an arbitrary changed caused an issue!

File: src/components/app.jsx (hunk 4 above)
Comment (briandangerflynn) [reply src/components/app.jsx:8]: I haven't worked with React hooks before, but it's one of the features I'm most interested in learning! I actually thought about trying it for the first time on this project, but then decided that maybe an interview project wasn't the best place to experiment. From what I've read, though, it seems like functional components with hooks should be the default over class components, as they require fewer lines of code to accomplish the same thing.

File: src/components/app.jsx (hunk 5 above)
Comment (briandangerflynn) [reply src/components/app.jsx:36]: I almost delayed turning this project in this morning to add error handling. Right now, if there's a network error on any of my HTTP requests, the user will have no idea that there was an error and think that the app just doesn't work. Not a great UX!

Normally, I would have wrapped my API calls in `try` / `catch`, added any error messages to state, and then rendered them in the UI. I actually began this process; however, some of the design choices I made earlier were problematic for this endeavor. For example, I stuffed both the current weather API call and the 5 day forecast API call into the same function calls. This made error handling sloppy. To fix it, I would have needed to separate them. 

If this were a real-world app, I definitely would have separated them, not only because of this messy error handling, but also because many users wouldn't need both the current weather and the 5 day forecast, so it forces the app to do more HTTP requests than necessary. I actually began this process for this project, then decided, since it was an extra feature, that maybe I should just turn it in and discuss with you. Hope that was the right choice!

**Comment by briandangerflynn** [comment]
Thanks for reviewing Jack! Looking forward to any additional thoughts. Much appreciated!
